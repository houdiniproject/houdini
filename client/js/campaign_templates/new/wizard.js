require('../../common/pikaday-timepicker')
require('../../components/wizard')
require('../../common/image_uploader')
var confirmation = require('../../common/confirmation')
var format_err = require('../../common/format_response_error')

appl.def('advance_campaign_template_name_step', function(form_obj) {
  appl.def('new_campaign_template', form_obj)
  appl.wizard.advance('new_campaign_template_wiz')
})

// Post a new campaign template
appl.def('create_campaign_template', function(el) {
	var form_data = utils.toFormData(appl.prev_elem(el))
	form_data = utils.mergeFormData(form_data, appl.new_campaign_template)
	appl.def('new_campaign_template_wiz.loading', true)

	post_campaign_template(form_data)
		.then(function(req) {
			appl.notify("Campaign template created!")
      var template_id = JSON.parse(req.response).id
			appl.redirect('/nonprofits/' + app.nonprofit_id + '/campaign_templates')
		})
		.catch(function(req) {
			appl.def('new_campaign_template_wiz.loading', false)
			appl.def('new_campaign_template_wiz.error', req.responseText)
		})
})


var Pikaday = require('pikaday')
var moment = require('moment')
new Pikaday({
	field: document.querySelector('.js-date-picker'),
	format: 'M/D/YYYY',
	minDate: moment().toDate()
})

// Using the bare-bones XMLHttpRequest API so we can post form data and upload the image
function post_campaign_template(form_data) {
	return new Promise(function(resolve, reject) {
		var req = new XMLHttpRequest()
		req.open("POST", '/nonprofits/' + app.nonprofit_id + '/campaign_templates')
		req.setRequestHeader('X-CSRF-Token', window._csrf)
		req.send(form_data)
		req.onload = function(ev) {
			if(req.status === 200) resolve(req)
			else reject(req)
		}
	})
}

appl.def('delete_template', function(id) {
  var result = confirmation('Are you sure?')
  result.confirmed = function() {
    appl.def('loading', true)
    var url = '/nonprofits/' + app.nonprofit_id + '/campaign_templates/' + id

    return new Promise(function(resolve, reject) {
      var req = new XMLHttpRequest()
      req.open("DELETE", url)
      req.setRequestHeader('X-CSRF-Token', window._csrf)
      req.send({ campaign_template: {id: id} })
      req.onload = function(ev) {
        if(req.status === 204) resolve(req)
        else reject(req)
      }
    }).then(function() {
      appl.def('loading', false)
      appl.notify('Successfully deleted template.')
      appl.redirect('/nonprofits/' + app.nonprofit_id + '/campaign_templates')
    })
  }
})


appl.def('create_campaign', function(campaign_params) {
  appl.def('loading', true)

  var url = '/nonprofits/' + app.nonprofit_id + '/campaigns'
  var params = new Object
  params.campaign = JSON.parse(campaign_params)
  params.campaign.profile_id = app.profile_id

  return new Promise(function(resolve, reject) {
    var req = new XMLHttpRequest()
    req.open("POST", url)
    req.setRequestHeader('X-CSRF-Token', window._csrf)
    req.setRequestHeader('Content-Type', 'application/json')
    req.send(JSON.stringify(params))
    req.onload = function(ev) {
      if(req.status === 200) resolve(req)
      else reject(req)
    }
  }).then(function(req) {
    appl.def('loading', false)
    appl.notify('Redirecting you to your campaignâ€¦')
    var campaign_id = JSON.parse(req.response).id
    appl.redirect(url + '/' + campaign_id)
  })
  .catch(function(req) {
    appl.def('loading', false)
    appl.notify(req.responseText)
  })
})
