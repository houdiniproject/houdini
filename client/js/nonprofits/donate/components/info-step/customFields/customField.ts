// License: LGPL-3.0-or-later
import { CustomFieldDescription } from "../../../types";
const h = require('snabbdom/h');


const typeToFormInputName = {
  'supporter': 'customFields'
}

export default function customField(field: CustomFieldDescription) : ReturnType<typeof h> {
  return h('input', {
    props: {
      name: `${typeToFormInputName[field.type]}[${field.name}]`,
      placeholder: field.label
    }
  });
}