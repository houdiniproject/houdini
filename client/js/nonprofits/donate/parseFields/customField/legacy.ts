// License: LGPL-3.0-or-later
import {map as Rmap, split as Rsplit, trim as Rtrim} from 'ramda';
import { CustomFieldDescription } from '.';

export function parseCustomField(f:string) :CustomFieldDescription {
  const [name, label] = Rmap(Rtrim, Rsplit(':', f))
  return {name, label: label || name, type: 'supporter'};
}