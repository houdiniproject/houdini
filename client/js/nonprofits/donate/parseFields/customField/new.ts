// License: LGPL-3.0-or-later
import { CustomFieldDescription } from ".";

export function parseCustomField(input:string) : CustomFieldDescription {
  const [name, ...rest] = input.split(":").map((s) => s.trim())
  const label = rest.length > 0 ? rest[0] : null;

  return {name, label: label || name, type: 'supporter' } as CustomFieldDescription;
};
