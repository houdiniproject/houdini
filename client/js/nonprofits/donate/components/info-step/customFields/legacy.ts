// License: LGPL-3.0-or-later
import {map as Rmap} from 'ramda';
import { CustomFieldDescription } from '../../../types';
import customField from './customField';
const h = require('snabbdom/h');

export function customFields(fields?:CustomFieldDescription[]|null): ReturnType<typeof h> | '' {
  if (!fields) return '';
  
  return h('div', Rmap(customField, fields));
}
