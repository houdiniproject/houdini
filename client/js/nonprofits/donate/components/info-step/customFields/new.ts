// License: LGPL-3.0-or-later
import { CustomFieldDescription } from '../../../types';
import customField from './customField';
const h = require('snabbdom/h');

export function customFields(fields?:CustomFieldDescription[]|null): ReturnType<typeof h> | '' {
  if (!fields) return '';
  
  return h('div', fields.map(customField));
}
