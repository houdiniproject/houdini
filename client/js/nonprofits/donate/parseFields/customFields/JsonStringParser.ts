// License: LGPL-3.0-or-later
import { parse } from 'json5';
import { CustomFieldDescription, isCustomFieldDescription } from '../customField';


export default class JsonStringParser {
  public errors:SyntaxError[] = [];
  public readonly results: CustomFieldDescription[] = [];
  constructor(public readonly fieldsString:string) {
    this._parse();
  }

  get isValid(): boolean {
    return this.errors.length == 0;
  }


  private _parse = (): void => {
    try {
      const result = parse(this.fieldsString)
      if (result instanceof Array) {
        result.forEach((i) => {
          if (isCustomFieldDescription(i)) {
            this.results.push({ ...i});
          }
          
          else {
            this.errors.push(new SyntaxError(JSON.stringify(i) + " is not a valid custom field description"))
          }
        });
      }
      else {
        this.errors.push(new SyntaxError("Input did not parse to an array"))
      }
    }
    catch(e:any) {
      this.errors.push(e)
    }
  }
}

