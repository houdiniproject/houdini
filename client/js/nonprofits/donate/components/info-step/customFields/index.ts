// License: LGPL-3.0-or-later
import {customFields} from './legacy';

export default customFields;
