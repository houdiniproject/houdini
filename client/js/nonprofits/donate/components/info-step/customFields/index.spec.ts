// License: LGPL-3.0-or-later
import customFields from "."
import {customFields as legacyCustomFields} from './legacy';
import {customFields as newCustomFields} from './new';



describe.each([
  ['with index', customFields],
  ['with legacy', legacyCustomFields],
  ['with new', newCustomFields]
])('customFields', (name, method) => {
  describe(name, () => {
    it('sets the the field correctly', () => {
      expect(method([
        {name: "Name of Field", label: "Field Label", type: 'supporter'},
        {name: "Another Field Name", label: "Label2", type: 'supporter'}
      ])).toMatchSnapshot();
    });
  
    it('returns blank string correctly with nothing passed', () => {
      expect(method()).toBe("");
    });
  
    it('returns blank string correctly with null passed', () => {
      expect(method(null)).toBe("");
    });
  });
});