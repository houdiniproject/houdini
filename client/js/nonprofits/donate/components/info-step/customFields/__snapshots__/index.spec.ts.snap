// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`customFields with index sets the the field correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "name": "customFields[Name of Field]",
          "placeholder": "Field Label",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "input",
      "text": undefined,
    },
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "name": "customFields[Another Field Name]",
          "placeholder": "Label2",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "input",
      "text": undefined,
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`customFields with legacy sets the the field correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "name": "customFields[Name of Field]",
          "placeholder": "Field Label",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "input",
      "text": undefined,
    },
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "name": "customFields[Another Field Name]",
          "placeholder": "Label2",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "input",
      "text": undefined,
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;

exports[`customFields with new sets the the field correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "name": "customFields[Name of Field]",
          "placeholder": "Field Label",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "input",
      "text": undefined,
    },
    Object {
      "children": undefined,
      "data": Object {
        "props": Object {
          "name": "customFields[Another Field Name]",
          "placeholder": "Label2",
        },
      },
      "elm": undefined,
      "key": undefined,
      "sel": "input",
      "text": undefined,
    },
  ],
  "data": Object {},
  "elm": undefined,
  "key": undefined,
  "sel": "div",
  "text": undefined,
}
`;
